src/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -66,27 +66,35 @@ newtype IndexOrdinary = IndexOrdinary (Vector SerialisedKey)
     if there is no such pair, the result is an arbitrary but valid page span.
 -}
 search :: SerialisedKey -> IndexOrdinary -> PageSpan
-search key (IndexOrdinary lastKeys)
-    = assert (length lastKeys > 0) $ PageSpan (PageNo start) (PageNo end)
-    where
-
-    start :: Int
-    start | protoStart < pageCount = protoStart
-          | otherwise              = maybe 0 succ $
-                                     findIndexR (/= lastKeys ! pred pageCount) $
-                                     lastKeys
-
-    end :: Int
-    end | protoStart < pageCount
-            = start + length (takeWhile (== key) (drop (succ start) lastKeys))
-        | otherwise
-            = pred pageCount
+search key (IndexOrdinary lastKeys) = assert (length lastKeys > 0) result where
 
     protoStart :: Int
-    protoStart = binarySearchL lastKeys key
+    !protoStart = binarySearchL lastKeys key
 
     pageCount :: Int
-    pageCount = length lastKeys
+    !pageCount = length lastKeys
+
+    result :: PageSpan
+    !result | protoStart < pageCount
+                = let
+
+                      end :: Int
+                      !end = maybe (pred pageCount) (+ protoStart) $
+                             findIndex (/= lastKeys ! protoStart)  $
+                             drop (succ protoStart) lastKeys
+
+                  in PageSpan (PageNo $ protoStart)
+                              (PageNo $ end)
```

Comment:
The `$` are not needed here (and similarly for `start` below`), so I would leave them out.

Reply:
Usually, I don’t write such code; in fact, it seems to be the first time that I did. That said, I did so for a reason: The definition of `result` is basically about computing the indexes of the start and the end page, and therefore I wanted these two pieces of data to be clearly stated. However, the occurrences of the data constructors `PageSpan` and `PageNo` are distracting. By separating the actual data from the additional data constructors using `$` and aligning the arguments of `PageSpan` vertically, I could make the actual data somewhat stick out.